docs/codegen-tour.md

```diff
@@ -0,0 +1,843 @@
+        cursor.finish(FoobMarker {
+            flags_byte_len,
+            versioned_value_byte_start,
+        })
+    }
+}
+```
+
+Let's walk through this. Firstly, the whole process is based around a 'cursor'
+type, which is simply a way of advancing through the input data on a
+field-by-field basis. Where we need to know the value of a field in order to
+validate subsequent fields, we read that field into a local variable.
+Additionally, values that we have to compute based on other fields are currently
+cached in the marker struct, although this is an implementation detail and may
+change. Let's walk through this code, field by field:
+
+- **version**: as this is marked with the `#[version]` attribute, we read the
+  value into a local variable, since we will need to know the version when
+  reading any versioned fields.
+- **some_val**: this is a simple value, and we do not need to know what it is,
+  only that it exists. We `advance` the cursor by the appropriate number of
+  bytes.
+- **other_val**: ditto. The compiler will be able to combine these two
+  `advances` into a single operation.
+- **flags_count**: This value is referenced in the `#[count]` attribute on the
+  following field, and so we bind it to a local variable.
+- **flags**: the `#[count]` attribute indicates that the length of this array is
+  stored in the `flags_count` field. We determine the array length by
+  multiplying that value by the size of the array member, and we advance the
+  cursor by that number of bytes.
+- **versioned_value**: this field is only available if the `version` field is >=
+  to `1` (this is specified via the `#[available]` attribute). We record the
+  current cursor position (as an `Option`, which will be `Some` only if the
+  version is compatible) and then we advance the cursor by the size of the
+  field's type.
+
+Finally, having finished with each field, we call the `finish` method on the
+cursor: this performs a final bounds check, and instantiates the table with the
+provided marker.
+
+> ***note***:
+>
+> The `FontRead` trait is currently doing a bit of a double duty: in the case of
+> tables, it is expected to perform a very minimal validation (essentially just
+> bounds checking) but in the case of records it serves as an actual parse
+> function, returning a concrete instance of the type. It is possible that these
+> two roles should be separated?
+
+#### <a id="versioned-tables"></a> versioned tables
+
+As hinted at above, for tables that are versioned (which have a version field,
+and which have more than one known version value we do not generate a distinct
+table per version; instead we generate a single table. For fields that are
+available on all versions of a table, we generate getters as usual. For fields
+that are only available on certain versions, we generate getters that return an
+`Option` type, which will be `Some` in the case where that field is present for
+the current version.
```

Comment:
I immediately wonder if it would be nice to access all the fields at a version as a group after *one* Option resolution. If I want to code against say COLRv1 I'd rather do one check then have access to all the fields, e.g.

```
if let Some(colrv1) = colr_table.v1_fields {
  // colrv1 has non-Option fields, I can use them all
}
```

Reply:
Yea, my sense here is that we should see how the ergonomics feel in practice, and we can revisit as needed.

This also gets complicated when you think of multiple versions. Does `color_table.v2_fields` also return `v1` fields?

I *could* imagine a world where we have a sort of 'upcasting' ability, where you can do, like:

```rust
if let Some(color_v1) = color_table.as_color_v1() {
    // color_v1 is a separate type that includes non-option getters
}
```

I think this pattern could be interesting, if we end up unsatisfied with the current ergonomics.